### PR TITLE
#59 Login

### PR DESCRIPTION
**Pull request by Vatavuk**
For #59, implemented login method. Left two puzzles for further development

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#156](https://github.com/timelesslounge/timelessis/issues/156) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/b4ec65d0-7c1c-4615-86ee-becde1b44c6a); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: tests/integration/it_test_auth.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    response = auth.login()
+    assert response.status_code == 200
+
+
+def test_forgot_password(client):
+    assert client.get("/auth/forgotpassword").status_code == 200
+
+
+def test_activate(client):
+    assert client.get("/auth/activate").status_code == 405
```
Comment (amihaiemil) [inline tests/integration/it_test_auth.py]: @Vatavuk Why should ``/activate`` return ``405``?

File: timeless/auth/auth.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+
+def login(username="", password=""):
+    """Login user"""
+    user = Employee.query.filter_by(username=username).first()
+    error = None
+    if user is None:
+        error = "Incorrect username."
+    elif not user.validate_password(password):
+        error = "Incorrect password."
```
Comment (amihaiemil) [inline timeless/auth/auth.py]: @Vatavuk The backend should return only error codes, not hardcoded messages. The codes should then be translated into (mapped to) Strings on the UI, so we can easily have internationalization or change error messages without having to rebuild + redeploy the backend.

File: timeless/auth/auth.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    if error is None:
+        session.clear()
+        session["user_id"] = user.id
+    return error
```
Comment (amihaiemil) [inline timeless/auth/auth.py]: @Vatavuk new line here

File: tests/test_auth.py [hunk 4]
```diff
@@ -1,12 +1,42 @@
...
+                            birth_date=datetime.utcnow(),
+                            registration_date=datetime.utcnow(),
+                            email="test@test.com", password="pass")
+        DB.session.add(employee)
+        DB.session.commit()
+        error = login("vgv", "pass")
+        DB.session.delete(employee)
+        DB.session.commit()
+        DB.session.remove()
+        assert ( error is None)
```
Comment (amihaiemil) [inline tests/test_auth.py]: @Vatavuk new line here

File: timeless/auth/auth.py (hunk 2 above)
Comment (Vatavuk) [reply timeless/auth/auth.py]: @amihaiemil I agree, but I believe that this should be done in the upper layer, in the view, not here.

File: tests/integration/it_test_auth.py (hunk 1 above)
Comment (Vatavuk) [reply tests/integration/it_test_auth.py]: @amihaiemil this is the old code, I haven't change it, I've just moved it in `integration` package.

File: timeless/auth/auth.py (hunk 2 above)
Comment (amihaiemil) [reply timeless/auth/auth.py]: @Vatavuk I think it would be fine to do return it here, but that's just my opinion. Yes, it's fine to leave a puzzle :)

**Comment by amihaiemil** [comment]
@Vatavuk the idea of enhancing the puzzle is fine (don't forget to update the PR's title too, please)

File: timeless/auth/auth.py (hunk 2 above)
Comment (Vatavuk) [reply timeless/auth/auth.py]: @amihaiemil I've misunderstood you, yes messages should be change to codes in this class. I've put a puzzle there.

**Comment by Vatavuk** [comment]
@amihaiemil fixed

**Comment by Vatavuk** [comment]
@amihaiemil why do I need to update PR title?

**Comment by amihaiemil** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@vatavuk just a single comment

**Comment by Vatavuk** [comment]
@emilianodellacasa fixed

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@vatavuk please check rultor

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by Vatavuk** [comment]
@emilianodellacasa will fix it, tests are executed before migrations

**Comment by Vatavuk** [comment]
@emilianodellacasa fixed, please try again

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#156](https://github.com/timelesslounge/timelessis/issues/156) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one major issue was found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by amihaiemil** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@amihaiemil thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**